sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/InsertAdaptiveSparkPlan.scala

```diff
@@ -0,0 +1,118 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.adaptive
+
+import scala.collection.concurrent.TrieMap
+import scala.collection.mutable
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.expressions
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution._
+import org.apache.spark.sql.execution.command.ExecutedCommandExec
+import org.apache.spark.sql.execution.exchange.{Exchange, ShuffleExchangeExec}
+import org.apache.spark.sql.types.StructType
+
+/**
+ * This rule wraps the query plan with an [[AdaptiveSparkPlanExec]], which executes the query plan
+ * and re-optimize the plan during execution based on runtime data statistics.
+ */
+case class InsertAdaptiveSparkPlan(session: SparkSession) extends Rule[SparkPlan] {
+
+  private val conf = session.sessionState.conf
+
+  // Exchange-reuse is shared across the entire query, including sub-queries.
+  private val stageCache = new TrieMap[StructType, mutable.Buffer[(Exchange, QueryStageExec)]]()
+
+  override def apply(plan: SparkPlan): SparkPlan = plan match {
+    case _: ExecutedCommandExec => plan
+    case _ if conf.runtimeReoptimizationEnabled
+      && supportAdaptive(plan) =>
+      try {
+        // Plan sub-queries recursively and pass in the shared stage cache for exchange reuse. Fall
+        // back to non-adaptive mode if adaptive execution is supported in any of the sub-queries.
+        val subqueryMap = planSubqueries(plan)
+        // Run preparation rules.
+        val preparations = AdaptiveSparkPlanExec.createQueryStagePreparationRules(
+          session.sessionState.conf, subqueryMap)
+        val newPlan = AdaptiveSparkPlanExec.applyPhysicalRules(plan, preparations)
+        logDebug(s"Adaptive execution enabled for plan: $plan")
+        AdaptiveSparkPlanExec(newPlan, session.cloneSession(), subqueryMap, stageCache)
+      } catch {
+        case _: SubqueryAdaptiveNotSupportedException =>
+          plan
+      }
+    case _ => plan
+  }
+
+  private def supportAdaptive(plan: SparkPlan): Boolean = {
+    sanityCheck(plan) &&
+      !plan.logicalLink.exists(_.isStreaming) &&
+      plan.children.forall(supportAdaptive)
+  }
+
+  private def sanityCheck(plan: SparkPlan): Boolean = plan match {
+    case _: SortExec | _: ShuffleExchangeExec => true
+    case _ => plan.logicalLink.isDefined
+  }
+
+  private def planSubqueries(plan: SparkPlan): Map[Long, ExecSubqueryExpression] = {
+    val subqueryMapBuilder = mutable.HashMap.empty[Long, ExecSubqueryExpression]
+    plan.foreach(_.expressions.foreach(_.foreach {
+      case expressions.ScalarSubquery(p, _, exprId)
+          if !subqueryMapBuilder.contains(exprId.id) =>
+        val executedPlan = getExecutedPlan(p)
+        val scalarSubquery = ScalarSubquery(
```

Comment:
Please use `execution. ScalarSubquery` here so people don't think you are inserting the same expression here.